### PR TITLE
feat: support env overrides for the adapter

### PR DESCRIPTION
**Pull request by tjohnson31415**
#### Motivation

We are working to align CRDs between this repo and kserve/kserve. In https://github.com/kserve/modelmesh-serving/pull/148, we mirrored updates to the ServingRuntime CRDs. This PR implements use of BuiltInAdapter.Env.

#### Modifications

- allow values in BuiltInAdapter.Env to override environment variables added by the controller

### CODE REVIEW

File: controllers/modelmesh/runtime_test.go [hunk 1]
```diff
@@ -73,30 +95,54 @@ func TestOverlayMockRuntime(t *testing.T) {
 				Spec: corev1.PodSpec{
 					Containers: []corev1.Container{
 						{
-							Name: "mm",
+							Name: adapterType,
```
Comment (pvaneck) [inline controllers/modelmesh/runtime_test.go:98]: nit: Not a big deal, but shouldn't this be maintained as "mm" to coincide with the actual initial container name?

File: controllers/modelmesh/runtime.go [hunk 2]
```diff
@@ -229,8 +231,13 @@ func (m *Deployment) addRuntimeToDeployment(deployment *appsv1.Deployment) error
...
 			},
-			{}, {}, // allocate larger array to avoid reallocation
-		}[:7]
+		}
+
+		// To avoid reallocation, count the maximum number of env vars there could be
+		numVarsDefault := len(defaultEnvVars) // count of default env vars
+		numVarsOptional := 2                  // count of envs that may be added from annotations below
+		builtInAdapterContainer.Env = make([]corev1.EnvVar, numVarsDefault, numVarsDefault+numVarsOptional+len(rt.Spec.BuiltInAdapter.Env))
+		copy(builtInAdapterContainer.Env, defaultEnvVars)
```
Comment (njhill) [inline controllers/modelmesh/runtime.go:240]: Not sure it's worth the extra complexity at this point, especially since this now does involve a separate array allocation. Maybe just revert the changes above here? Feel free to also remove the existing larger-array alloc logic if you don't like it.

File: controllers/modelmesh/runtime.go [hunk 3]
```diff
@@ -187,15 +187,17 @@ func (m *Deployment) addRuntimeToDeployment(deployment *appsv1.Deployment) error
 
 		// the puller and adapter containers are the same image and are given the
 		// same resources
-		builtInAdapterContainer.Resources = *m.PullerResources
+		if m.PullerResources != nil {
```
Comment (njhill) [inline controllers/modelmesh/runtime.go:190]: I don't think this should ever be nil here, might be better to let it fail with deref error since it would imply a bug. Or otherwise at least add an error log in the nil case?

**Review by njhill — COMMENTED** [review]
Thanks @tjohnson31415 looks good just minor comments

File: controllers/modelmesh/runtime.go (hunk 3 above)
Comment (tjohnson31415) [reply controllers/modelmesh/runtime.go:190]: The main reason for the change is to support the test changes. Adding use of `BuiltInAdapter` to the tests triggered a nil pointer exception here where the Deployment is not created with all fields being set. I will add a test helper to create a complete `modelmesh.Deployment` for the unit tests (including the test logger I added).

Also I know the controller doesn't support it, but it is valid for a container to not have any resources specified (BestEffort QoS). If it was supported, it would be useful for small dev/test environments

File: controllers/modelmesh/runtime_test.go (hunk 1 above)
Comment (tjohnson31415) [reply controllers/modelmesh/runtime_test.go:98]: Ah, I think I changed this in the wrong spot... With the addition of BuiltInAdapter, there should be a container with the same name as the `ServerType` (though this is checked well before the real code path would reach this point), so the runtime container is the thing that should have a matching name.

Nice catch.

**Review by pvaneck — COMMENTED** [review]
/lgtm

**Review by njhill — APPROVED** [review]
